code/modules/tgui/tgui_alert.dm

```diff
@@ -0,0 +1,169 @@
+ * Arguments:
+ * * user - The user to show the alert to.
+ * * message - The content of the alert, shown in the body of the TGUI window.
+ * * title - The of the alert modal, shown on the top of the TGUI window.
+ * * buttons - The options that can be chosen by the user, each string is assigned a button on the UI.
+ * * timeout - The timeout of the alert, after which the modal will close and qdel itself. Set to zero for no timeout.
+ */
+/proc/tgui_alert(mob/user, message, title, list/buttons, timeout = 60 SECONDS)
+	if (!user)
+		user = usr
+	if (!istype(user))
+		if (istype(user, /client))
+			var/client/client = user
+			user = client.mob
+		else
+			return
+	var/datum/tgui_modal/alert = new(user, message, title, buttons, timeout)
+	alert.ui_interact(user)
+	alert.wait()
+	if (alert)
+		. = alert.choice
+		qdel(alert)
+
+/**
+ * Creates an asynchronous TGUI alert window with an associated callback.
+ *
+ * This proc should be used to create alerts that invoke a callback with the user's chosen option.
+ * Arguments:
+ * * user - The user to show the alert to.
+ * * message - The content of the alert, shown in the body of the TGUI window.
+ * * title - The of the alert modal, shown on the top of the TGUI window.
+ * * buttons - The options that can be chosen by the user, each string is assigned a button on the UI.
+ * * callback - The callback to be invoked when a choice is made.
+ * * timeout - The timeout of the alert, after which the modal will close and qdel itself. Set to zero for no timeout.
+ */
+/proc/tgui_alert_async(mob/user, message, title, list/buttons, datum/callback/callback, timeout = 60 SECONDS)
+	if (!user)
+		user = usr
+	if (!istype(user))
+		if (istype(user, /client))
+			var/client/client = user
+			user = client.mob
+		else
+			return
+	var/datum/tgui_modal/async/alert = new(user, message, title, buttons, callback, timeout)
+	alert.ui_interact(user)
+
+/**
+ * # tgui_modal
+ *
+ * Datum used for instantiating and using a TGUI-controlled modal that prompts the user with
+ * a message and has buttons for responses.
+ */
+/datum/tgui_modal
+	/// The title of the TGUI window
+	var/title
+	/// The textual body of the TGUI window
+	var/message
+	/// The list of buttons (responses) provided on the TGUI window
+	var/list/buttons
+	/// The button that the user has pressed, null if no selection has been made
+	var/choice
+	/// The time at which the tgui_modal was created, for displaying timeout progress.
+	var/start_time
+	/// The lifespan of the tgui_modal, after which the window will close and delete itself.
+	var/timeout
+	/// Boolean field describing if the tgui_modal was closed by the user.
+	var/closed
+
+/datum/tgui_modal/New(mob/user, message, title, list/buttons, timeout)
+	src.title = title
+	src.message = message
+	src.buttons = buttons.Copy()
+	if (timeout)
+		src.timeout = timeout
+		start_time = world.time
+		SPAWN_DBG(timeout)
+			qdel(src)
+			. = ..()
```

Comment:
Sanity check on this being the correct indent level here.

Reply:
I dunno, tg didn't have it, but our linter required it, so that's where i thought it should go. Could be an indent up though, yeah.

Reply:
@pali6 ?

Reply:
Incorrect indentation level, yeah. 
```suggestion
		. = ..()
```